### ISSUE TITLE
airodump-ng: sometimes show wrong channel when network not have a valid essid

### ISSUE THREAD

**Issue by aircrack-ng** [aircrack-ng/aircrack-ng#670]
**Reported by m4tr1x on 8 Sep 2009 00:16 UTC**

hi,

sometimes airodump-ng show wrong channel when it found a network with hidden/invalid essid (ex.: <length:  0>).

This bug it's related to bad radiotap parsing,

the result it's a wrong converted value (133, 108, etc...)

Good work

**Comment by aircrack-ng** [comment]
**Comment by misterx on 8 Sep 2009 00:16 UTC**

See also #647.

**Comment by aircrack-ng** [comment]
**Comment by m4tr1x on 8 Sep 2009 00:16 UTC**

if you convert the bad value with this macro:

```
#define EXTRACT_16BITS(p) \
                      ((u_int16_t)((u_int16_t)*((const u_int8_t *)(p) + 1) << 8 | \
                      (u_int16_t)*((const u_int8_t *)(p) + 0)))
```
the result is a valid 802.11b/g channel

**Comment by aircrack-ng** [comment]
**Comment by someone on 8 Sep 2009 00:16 UTC**

It might be in osdep/byteorder.h: 
```
#define ___my_swab16(x)
```

**Comment by aircrack-ng** [comment]
**Modified by misterx on 8 Sep 2009 00:16 UTC**

**Comment by aircrack-ng** [comment]
**Comment by misterx on 8 Sep 2009 00:16 UTC**

That's a known issue, I have been able to replicate it and I have a capture file to replicate it. 

Duplicate, see tickets #689, #714 and #535

**Comment by aircrack-ng** [comment]
**Comment by frankwoelk on 8 Sep 2009 00:16 UTC**

Netgear-Router this time:
BSSID, First time seen, Last time seen, channel, Speed, Privacy, Cipher, Authentication, Power, # beacons, # IV, LAN IP, ID-length, ESSID, Key

00:09:5B:C7:99:88, 2010-04-28 22:03:46, 2010-04-28 22:15:38, 158,  -1,     ,       ,   ,  -1,        0,        0,   0.  0.  0.  0,   0, , 



Station MAC, First time seen, Last time seen, Power, # packets, BSSID, Probed ESSIDs

00:22:68:B1:58:3C, 2010-04-28 21:47:25, 2010-04-28 22:17:43, -104,       14, 00:09:5B:C7:99:88, NETGEAR

My network driver is ath9k ubuntu 9.10 kernel 2.6.31-20-generic

**Comment by aircrack-ng** [comment]
**Comment by anonymous on 8 Sep 2009 00:16 UTC**

Hi. This ticket hasn't seen any activity for 6 months and as far as I can tell the issue is still there (certain ssid's aren't handled well and as a result airodump-ng gives meaningless channel numbers: e.g. 133).

Is a fix in the pipeline?

**Comment by aircrack-ng** [comment]
**Comment by nebuchadnezzarrk on 8 Sep 2009 00:16 UTC**

Yes I could really use a solution as well.  I have searched high and low and can't seem to fix the issue.  Anyone else have any luck finding a resolution?

**Comment by aircrack-ng** [comment]
**Comment by eabarcaortiz on 8 Sep 2009 00:16 UTC**

Hi.

   If I understood it well, the "#define EXTRACT_16BITS(p)" is a fix and it works...

   But I tried to 'apply' it to the "osdep/byteorder.h" but no luck.

   Changing "#define _my_swab16(x)" by EXTRACT..., just the contents of the macro, changing also parameter x -> p... nothing works. Also using the macro with EXTRACT_16BITS(ap_cur->channel) has no good result (seg fault, probably I need to learn C again...) :)

   Can somebody explain the fix a little bit more or supply a diff?

Thanks!!

**Comment by aircrack-ng** [comment]
**Comment by anonymous on 8 Sep 2009 00:16 UTC**

Hi. Is there any hope of a patch to solve this?

Cheers.

**Comment by aircrack-ng** [comment]
**Comment by martin tee77 on 8 Sep 2009 00:16 UTC**

Issue is still apparent.

**Comment by aircrack-ng** [comment]
**Comment by anonymous on 8 Sep 2009 00:16 UTC**

Still have this issue. this bug is two years old. heh.

**Comment by aircrack-ng** [comment]
**Comment by anonymous on 8 Sep 2009 00:16 UTC**

I still have this bug too. Is any developer reading this?

**Comment by aircrack-ng** [comment]
**Comment by misterx on 8 Sep 2009 00:16 UTC**

Yes.

**Comment by aircrack-ng** [comment]
**Comment by BobWilkins on 8 Sep 2009 00:16 UTC**

Apart for reading this, is any developer working on it?
1.1 here.

**Comment by aircrack-ng** [comment]
**Comment by doda on 8 Sep 2009 00:16 UTC**

It would be great if someone could take a look at this.

**Comment by aircrack-ng** [comment]
**Comment by anonymous on 8 Sep 2009 00:16 UTC**

would like to see this too...

**Comment by aircrack-ng** [comment]
**Comment by anonymous on 8 Sep 2009 00:16 UTC**

is there a fix available for this error or not?

**Comment by aircrack-ng** [comment]
**Comment by anonymous on 8 Sep 2009 00:16 UTC**

Hi, running 1.1-4 on linux mint. Issue still present.

**Comment by aircrack-ng** [comment]
**Comment by anonymous on 8 Sep 2009 00:16 UTC**

Still here in build from SVN version 2333

**Comment by aircrack-ng** [comment]
**Comment by anonymous on 8 Sep 2009 00:16 UTC**

Still here!!!! 3 years!!! pls fix this

**Comment by aircrack-ng** [comment]
**Comment by anonymous on 8 Sep 2009 00:16 UTC**

I would like to confirm this bug is still present.

**Comment by aircrack-ng** [comment]
**Comment by anonymous on 8 Sep 2009 00:16 UTC**

got it in kali linux this morning

**Comment by aircrack-ng** [comment]
**Comment by anonymous on 8 Sep 2009 00:16 UTC**

Getting the issue in Kali.
Says the hidden ESSIS (length: 0) has a Ch of 133.

**Comment by aircrack-ng** [comment]
**Comment by misterx on 8 Sep 2009 00:16 UTC**

See also #647

**Comment by aircrack-ng** [comment]
**Comment by anonymous on 8 Sep 2009 00:16 UTC**

This is is due to wrong radiotap header parsing.

Be warned: When the channel is taken from radiotap header, it is NOT always the right one as packets could also be received from neighbour channels! Maybe we should implement an indicator for that.

Fix (src/osdep/linux.c):

```
Index: linux.c
===================================================================
--- linux.c	(Revision 2371)
+++ linux.c	(Arbeitskopie)
@@ -739,7 +739,7 @@
                 break;
 
             case IEEE80211_RADIOTAP_CHANNEL:
-                ri->ri_channel = *iterator.this_arg;
+                ri->ri_channel = getChannelFromFrequency(le16toh(*(uint16_t*)iterator.this_arg));
                 got_channel = 1;
                 break;
```

 

But I also recomment to update to latest radiotap from git http://git.sipsolutions.net/?p=radiotap.git. Just check it out into osdep (delete old radiotap folder beforehand) and apply my attached patch

**Comment by aircrack-ng** [comment]
**Comment by misterx on 8 Sep 2009 00:16 UTC**

Fixed in aircrack-ng/aircrack-ng@5f23bdad0e9d640ae87ca35edb20d3ba90ace81b.
